primitive-types/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Primitive types shared by Substrate and Parity Ethereum.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "impl-serde", not(feature = "std")))]
+compile_error!("Feature \"impl-serde\" requires feature \"std\" to build.");
+
+#[cfg(all(feature = "impl-rlp", not(feature = "std")))]
+compile_error!("Feature \"impl-rlp\" requires feature \"std\" to build.");
+
+#[macro_use]
+extern crate uint;
+
+#[macro_use]
+extern crate fixed_hash;
+
+#[macro_use]
+extern crate crunchy;
+
+#[cfg(feature = "impl-serde")]
+#[macro_use]
+extern crate impl_serde;
+
+#[cfg(feature = "impl-codec")]
+#[macro_use]
+extern crate impl_codec;
+
+#[cfg(feature = "impl-rlp")]
+#[macro_use]
+extern crate impl_rlp;
+
+construct_uint!(U256, 4);
+#[cfg(feature = "impl-serde")] impl_uint_serde!(U256, 4);
+#[cfg(feature = "impl-codec")] impl_uint_codec!(U256, 4);
+#[cfg(feature = "impl-rlp")] impl_uint_rlp!(U256, 4);
+
+impl U256 {
+	/// Multiplies two 256-bit integers to produce full 512-bit integer
+	/// No overflow possible
+	#[inline(always)]
+	pub fn full_mul(self, other: U256) -> U512 {
+		U512(uint_full_mul_reg!(U256, 4, self, other))
+	}
+}
+
+impl From<U256> for U512 {
+	fn from(value: U256) -> U512 {
+		let U256(ref arr) = value;
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U512(ret)
+	}
+}
+
+impl From<U512> for U256 {
+	fn from(value: U512) -> U256 {
+		let U512(ref arr) = value;
+		if arr[4] | arr[5] | arr[6] | arr[7] != 0 {
+			panic!("From<U512> for U256: encountered overflow")
+		}
+		let mut ret = [0; 4];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U256(ret)
+	}
+}
+
+impl<'a> From<&'a U256> for U512 {
+	fn from(value: &'a U256) -> U512 {
+		let U256(ref arr) = *value;
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U512(ret)
+	}
+}
+
+impl<'a> From<&'a U512> for U256 {
+	fn from(value: &'a U512) -> U256 {
+		let U512(ref arr) = *value;
+		if arr[4] | arr[5] | arr[6] | arr[7] != 0 {
+			panic!("From<&U512> for U256: encountered overflow")
+		}
+		let mut ret = [0; 4];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U256(ret)
+	}
+}
+
+construct_uint!(U512, 8);
+#[cfg(feature = "impl-serde")] impl_uint_serde!(U512, 8);
+#[cfg(feature = "impl-codec")] impl_uint_codec!(U512, 8);
+#[cfg(feature = "impl-rlp")] impl_uint_rlp!(U512, 8);
+
+construct_fixed_hash! {
+	/// Fixed-size uninterpreted hash type with 20 bytes (160 bits) size.
+	pub struct H160(20);
+}
+#[cfg(feature = "impl-serde")] impl_fixed_hash_serde!(H160, 20);
+#[cfg(feature = "impl-codec")] impl_fixed_hash_codec!(H160, 20);
+#[cfg(feature = "impl-rlp")] impl_fixed_hash_rlp!(H160, 20);
+
+#[cfg(feature = "std")]
+impl From<u64> for H160 {
```

Comment:
@Robbepop maybe we should have this in `uint`?

Reply:
This is problematic since in `uint` you can also define `uint` that are smaller than `u64` and thus making this conversion failing. However, `From` impls are guaranteed to never fail. So we would need to make this conversion lossy in order to guarantee that it never fails but I am not sure if this is explicitly wanted behaviour since this may introduce unwanted semantics.

Reply:
This from impl is fairly short so I think it's okay to just be there.

Reply:
Hmm I'm thinking -- maybe we should remove those `From<u64>` conversions? AFAIK it's only used in Substrate, but the issue with it is that it assumes big-endian encoding. To fix those in Substrate is trivial -- we just replace any `From<u64>` by the explicit function `from_low_u64_be`.

Reply:
> replace any `From<u64>` by the explicit function `from_low_u64_b`

I like this solution. Let's do that. :)